Support globally defined shortcodes

eigengrau opened the pull request:
This adds support for a new config block, «global». Shortcodes defined in this
block are always applied, even when the jinja2 filter is not used. In case the
jinja2 filter is used, shortcodes from this section are applied in addition to
the shortcodes specified by the filter.

eigengrau committed:
Support globally defined shortcodes

This adds support for a new config block, «global». Shortcodes defined in this
block are always applied, even when the jinja2 filter is not used. In case the
jinja2 filter is used, shortcodes from this section are applied in addition to
the shortcodes specified by the filter.

eigengrau commented:
I wasn’t sure if requiring a Jinja filter to process the shortcodes was an explicit design choice. In any case I figured it might be useful to be able to use shortcodes without any changes to the template, too. To maintain backwards-compatibility, the section is called «global».

eigengrau commented:
Note that I didn’t bump the version yet, since I was making another PR.

skorokithakis commented:
Thanks for this, it's definitely useful. Using a filter wasn't so much a design choice as a necessity because messing with mistune's lexer would take more time than I had. Your PR is basically how shortcodes was always meant to work, thank you!

skorokithakis commented:
One question: What does the lexer do if a shortcode is encountered and no global section (and no filter) are defined? I don't want to step on any toes, so in that case it should leave the parsed text as it found it.

eigengrau commented:
Good catch. I think this should address this issue.

skorokithakis commented on lektor_shortcodes/__init__.py:
```diff
@@ -34,5 +78,9 @@ class ShortcodesPlugin(Plugin):
     description = u'Shortcodes for Lektor.'
 
+    def on_markdown_config(self, config):
+        lexer = ShortcodeLexer(self.get_config())
```
Would it not be easier to just have a ```if "global" not in sections: return``` here? That way we'd just not add the lexer if the global section was missing.

eigengrau replied:
Indeed! Amended… I’m using `section_as_dict` for the containment check though, because `config.keys()` yields an iterator in Python 3.

eigengrau committed:
Make global processing opt-in

We want to be relaxed about undefined shortcodes if the user didn’t actually
configure any global shortcodes. Until the global section is added to the
config, we don’t install the lexer.

skorokithakis commented:
Looks good to me! Hopefully `section_as_dict` returns `{}` if the section doesn't exist, which will do what we want. Thanks!

eigengrau commented:
> Hopefully section_as_dict returns {} if the section doesn't exist

I checked that before committing, and it did, so hopefully nothing will blow up there at least. ;-)

Thank you for making lektor-shortcodes!

skorokithakis commented:
Yep, I checked a bit and released 0.2.0. Thank you for contributing!
